server/src/main/java/org/opensearch/search/query/rewriters/BooleanFlatteningRewriter.java

```diff
@@ -0,0 +1,233 @@
+                return true;
+            }
+        }
+        for (QueryBuilder clause : boolQuery.filter()) {
+            if (clause instanceof BoolQueryBuilder && needsFlattening((BoolQueryBuilder) clause)) {
+                return true;
+            }
+        }
+        for (QueryBuilder clause : boolQuery.should()) {
+            if (clause instanceof BoolQueryBuilder && needsFlattening((BoolQueryBuilder) clause)) {
+                return true;
+            }
+        }
+        for (QueryBuilder clause : boolQuery.mustNot()) {
+            if (clause instanceof BoolQueryBuilder && needsFlattening((BoolQueryBuilder) clause)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private BoolQueryBuilder flattenBoolQuery(BoolQueryBuilder original) {
+        BoolQueryBuilder flattened = new BoolQueryBuilder();
+
+        flattened.boost(original.boost());
+        flattened.queryName(original.queryName());
+        flattened.minimumShouldMatch(original.minimumShouldMatch());
+        flattened.adjustPureNegative(original.adjustPureNegative());
+
+        flattenClauses(original.must(), flattened, ClauseType.MUST);
+        flattenClauses(original.filter(), flattened, ClauseType.FILTER);
+        flattenClauses(original.should(), flattened, ClauseType.SHOULD);
+        flattenClauses(original.mustNot(), flattened, ClauseType.MUST_NOT);
+
+        return flattened;
+    }
+
+    private void flattenClauses(List<QueryBuilder> clauses, BoolQueryBuilder target, ClauseType clauseType) {
+        for (QueryBuilder clause : clauses) {
+            if (clause instanceof BoolQueryBuilder) {
+                BoolQueryBuilder nestedBool = (BoolQueryBuilder) clause;
+
+                if (canFlatten(nestedBool, clauseType)) {
+                    // Flatten the nested bool query by extracting its clauses
+                    List<QueryBuilder> nestedClauses = getClausesForType(nestedBool, clauseType);
+                    for (QueryBuilder nestedClause : nestedClauses) {
+                        // Recursively flatten if needed
+                        if (nestedClause instanceof BoolQueryBuilder) {
+                            nestedClause = flattenBoolQuery((BoolQueryBuilder) nestedClause);
+                        }
+                        addClauseBasedOnType(target, nestedClause, clauseType);
+                    }
+                } else {
+                    // Can't flatten this bool, but recursively flatten its contents
+                    BoolQueryBuilder flattenedNested = flattenBoolQuery(nestedBool);
+                    addClauseBasedOnType(target, flattenedNested, clauseType);
+                }
+            } else {
+                // Non-boolean clause, add as-is
+                addClauseBasedOnType(target, clause, clauseType);
+            }
+        }
+    }
+
+    private boolean canFlatten(BoolQueryBuilder nestedBool, ClauseType parentType) {
```

Comment:
I'm wondering if FunctionScore query will be considered to be flatten here? Ideally, we should just skip them.